tracer/src/Datadog.Trace/AppSec/Coordinator/SecurityCoordinator.cs

```diff
@@ -96,6 +92,98 @@ internal readonly partial struct SecurityCoordinator
 
+    public IResult? RunWafForUser(string? userId = null, string? userLogin = null, string? userSessionId = null, bool fromSdk = false, Dictionary<string, string>? otherTags = null)
+    {
+        if (_httpTransport.IsBlocked)
+        {
+            return null;
+        }
+
+        IResult? result = null;
+        Dictionary<string, object>? addresses = null;
+        try
+        {
+            var additiveContext = GetOrCreateAdditiveContext();
+            if (additiveContext?.ShouldRunWith(_security, userId, userLogin, userSessionId, fromSdk) is { Count: > 0 } userAddresses)
+            {
+                addresses = userAddresses.ToDictionary(k => k.Key, object (v) => v.Value);
+
+                if (otherTags is not null)
+                {
+                    foreach (var kvp in otherTags)
+                    {
+                        if (!addresses.ContainsKey(kvp.Key))
+                        {
+                            addresses.Add(kvp.Key, kvp.Value);
+                        }
+                    }
+                }
+
+                SecurityReporter.LogAddressIfDebugEnabled(addresses);
+
+                // run the WAF and execute the results
+                result = additiveContext.Run(addresses, _security.Settings.WafTimeoutMicroSeconds);
+                additiveContext.CommitUserRuns(userAddresses, fromSdk);
+                RecordTelemetry(result);
+
+                if (_localRootSpan.Context.TraceContext is not null)
+                {
+                    _localRootSpan.Context.TraceContext.WafExecuted = true;
+                }
+            }
+        }
+        catch (Exception ex) when (ex is not BlockException)
+        {
+            if (addresses is not null)
+            {
+                var stringBuilder = StringBuilderCache.Acquire();
+                foreach (var kvp in addresses)
+                {
+                    stringBuilder.Append($"Key: {kvp.Key} Value: {kvp.Value}, ");
+                }
+
+                Log.Error(ex, "Call into the security module failed with arguments {Args}", StringBuilderCache.GetStringAndRelease(stringBuilder));
+            }
+        }
+
+        return result;
+    }
+
+    private static void RecordTelemetry(IResult? result)
+    {
+        if (result == null)
+        {
+            return;
+        }
+
+        if (result.Timeout)
+        {
+            TelemetryFactory.Metrics.RecordCountWafRequests(MetricTags.WafAnalysis.WafTimeout);
+        }
+        else if (result.ShouldBlock)
+        {
+            TelemetryFactory.Metrics.RecordCountWafRequests(MetricTags.WafAnalysis.RuleTriggeredAndBlocked);
+        }
+        else if (result.ShouldReportSecurityResult)
+        {
+            TelemetryFactory.Metrics.RecordCountWafRequests(MetricTags.WafAnalysis.RuleTriggered);
+        }
+        else
+        {
+            TelemetryFactory.Metrics.RecordCountWafRequests(MetricTags.WafAnalysis.Normal);
+        }
```

Comment:
nit, I think pattern matching makes this easier to read? 

```suggestion
        var metric = result switch
        {
            { Timeout: true } => MetricTags.WafAnalysis.WafTimeout,
            { ShouldBlock: true } => MetricTags.WafAnalysis.RuleTriggeredAndBlocked,
            { ShouldReportSecurityResult: true } => MetricTags.WafAnalysis.RuleTriggered,
            _ => MetricTags.WafAnalysis.Normal,
        };

        TelemetryFactory.Metrics.RecordCountWafRequests(metric);
```

Reply:
I did, it just slipped out 😅  but [e1fdfb5](https://github.com/DataDog/dd-trace-dotnet/pull/6431/commits/e1fdfb5bca961029bd6625b0a64922de8431837c) thanks